src/expression.c

```diff
@@ -6104,8 +6104,9 @@ Expression *TypeidExp::semantic(Scope *sc)
     {
         /* Get the dynamic type, which is .classinfo
          */
-        e = new DotIdExp(ea->loc, ea, Id::classinfo);
-        e = e->semantic(sc);
+        ea = ea->semantic(sc);
+        e = new TypeidExp(ea->loc, ea);
+        e->type = Type::typeinfoclass->type;
```

Comment:
I opened a new [bugzilla issue 14612](https://issues.dlang.org/show_bug.cgi?id=14612) for that non-intuitive behavior.

If an `InterfaceDeclaration *id` exists, there are two `TypeInfo`s - `id->csym` == `TypeInfo_Class` and `id->type->vtinfo` == `TypeInfo_Interface`. `typeid(i)` and `i.classinfo` return the former, but `typeid(I)` an d `typeid(J)` return the latter.

If we want to fix that, we would need to fix both dmd and druntime.